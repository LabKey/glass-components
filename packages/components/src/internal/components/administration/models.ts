interface Member {
    id: number;
    name: string;
    type: string;
}

interface Group {
    groupName: string;
    members: Member[];
}

export interface GroupMembership {
    [key: string]: Group;
}
