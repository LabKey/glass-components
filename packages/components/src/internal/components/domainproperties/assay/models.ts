/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { List, Record as ImmutableRecord } from 'immutable';

import { getServerContext, Utils } from '@labkey/api';

import { DomainDesign, FieldErrors } from '../models';
import { AppURL } from '../../../url/AppURL';
import { getAppHomeFolderPath } from '../../../app/utils';
import { Container } from '../../base/models/Container';

// See ExpProtocol.Status in 'platform' repository.
export enum Status {
    Active = 'Active',
    Archived = 'Archived',
}

export interface ProtocolTransformScript {
    runOnEdit: boolean;
    runOnImport: boolean;
    scriptPath: string;
}

export class AssayProtocolModel extends ImmutableRecord({
    allowBackgroundUpload: false,
    allowEditableResults: false,
    allowQCStates: false,
    allowTransformationScript: false,
    allowPlateMetadata: false,
    autoCopyTargetContainer: undefined,
    autoCopyTargetContainerId: undefined,
    autoLinkCategory: undefined,
    availableDetectionMethods: undefined,
    availableMetadataInputFormats: undefined,
    availablePlateTemplates: undefined,
    backgroundUpload: false,
    description: undefined,
    domains: undefined,
    editableResults: false,
    editableRuns: false,
    exception: undefined,
    metadataInputFormatHelp: undefined,
    moduleTransformScripts: undefined,
    name: undefined,
    plateMetadata: undefined,
    protocolId: undefined,
    protocolParameters: undefined,
    protocolTransformScripts: undefined,
    providerName: undefined,
    saveScriptFiles: false,
    selectedDetectionMethod: undefined,
    selectedMetadataInputFormat: undefined,
    selectedPlateTemplate: undefined,
    status: Status.Active,
    qcEnabled: undefined,
    excludedContainerIds: undefined,
}) {
    declare allowBackgroundUpload: boolean;
    declare allowEditableResults: boolean;
    declare allowQCStates: boolean;
    declare allowTransformationScript: boolean;
    declare allowPlateMetadata: boolean;
    declare autoCopyTargetContainer: {};
    declare autoCopyTargetContainerId: string;
    declare autoLinkCategory: string;
    declare availableDetectionMethods: [];
    declare availableMetadataInputFormats: {};
    declare availablePlateTemplates: [];
    declare backgroundUpload: boolean;
    declare description: string;
    declare domains: List<DomainDesign>;
    declare editableResults: boolean;
    declare editableRuns: boolean;
    declare exception: string;
    declare metadataInputFormatHelp: any;
    declare moduleTransformScripts: List<string>;
    declare name: string;
    declare plateMetadata: boolean;
    declare protocolId: number;
    declare protocolParameters: any;
    declare protocolTransformScripts: List<ProtocolTransformScript>;
    declare providerName: string;
    declare saveScriptFiles: boolean;
    declare selectedDetectionMethod: string;
    declare selectedMetadataInputFormat: string;
    declare selectedPlateTemplate: string;
    declare status: Status;
    declare qcEnabled: boolean;
    declare excludedContainerIds: string[];

    static create(raw: any): AssayProtocolModel {
        let domains = raw.domains || List<DomainDesign>();
        if (raw.domains && Utils.isArray(raw.domains)) {
            domains = List<DomainDesign>(
                raw.domains.map(domain => {
                    return DomainDesign.create(domain);
                })
            );
        }

        if (raw.protocolTransformScripts && Utils.isArray(raw.protocolTransformScripts)) {
            raw.protocolTransformScripts = List(raw.protocolTransformScripts);
        }
        if (raw.moduleTransformScripts && Utils.isArray(raw.moduleTransformScripts)) {
            raw.moduleTransformScripts = List<string>(raw.moduleTransformScripts);
        }

        // if this is not an existing assay, clear the name property so the user must set it
        const name = !raw.protocolId ? undefined : raw.name;

        // Issue 38685: for new assays, pre-select some required assay properties
        const model = new AssayProtocolModel({ ...raw, name, domains });
        if (model.isNew()) {
            if (model.allowDetectionMethodSelection() && model.availableDetectionMethods.length > 0) {
                raw.selectedDetectionMethod = List<string>(model.availableDetectionMethods).get(0);
            }
            if (model.allowPlateTemplateSelection() && model.availablePlateTemplates.length > 0) {
                raw.selectedPlateTemplate = List<string>(model.availablePlateTemplates).get(0);
            }
        }

        return new AssayProtocolModel({ ...raw, name, domains });
    }

    static serialize(model: AssayProtocolModel): any {
        // need to serialize the DomainDesign objects to remove the unrecognized fields
        const domains = model.domains.map(domain => {
            return DomainDesign.serialize(domain);
        });

        const json = model.merge({ domains }).toJS();

        // only need to serialize the id and not the autoCopyTargetContainer object
        delete json.autoCopyTargetContainer;
        delete json.exception;
        return json;
    }

    isActive(): boolean {
        return this.status === Status.Active;
    }

    getDomainByNameSuffix(name: string): DomainDesign {
        if (this.domains.size > 0) {
            return this.domains.find(domain => {
                return domain.isNameSuffixMatch(name);
            });
        }
    }

    get container(): string {
        const container = new Container(getServerContext().container);
        const domainContainerId = this.domainContainerId;

        return this.isNew()
            ? getAppHomeFolderPath(container)
            : domainContainerId === container.id
              ? container.path
              : domainContainerId;
    }

    get domainContainerId(): string {
        return this.getIn(['domains', 0, 'container']);
    }

    isNew(): boolean {
        return !this.protocolId;
    }

    allowPlateTemplateSelection(): boolean {
        return this.availablePlateTemplates && Utils.isArray(this.availablePlateTemplates);
    }

    allowDetectionMethodSelection(): boolean {
        return this.availableDetectionMethods && Utils.isArray(this.availableDetectionMethods);
    }

    allowMetadataInputFormatSelection(): boolean {
        return (
            this.availableMetadataInputFormats &&
            Utils.isObject(this.availableMetadataInputFormats) &&
            !Utils.isEmptyObj(this.availableMetadataInputFormats)
        );
    }

    validateTransformScripts(): string {
        if (this.protocolTransformScripts === undefined || this.protocolTransformScripts.size === 0) {
            return undefined;
        }

        // make sure we don't have any script inputs that are empty strings
        const hasEmptyScript = this.protocolTransformScripts.some(
            config => config?.scriptPath === undefined || config.scriptPath === null || config.scriptPath.length === 0
        );
        if (hasEmptyScript) {
            return 'Missing required transform script path.';
        }
    }

    getFirstDomainFieldError(): FieldErrors {
        const firstErrantDomain = this.domains.find(domain => domain.hasInvalidFields());
        return firstErrantDomain !== undefined ? firstErrantDomain.getInvalidFields().first().getErrors() : undefined;
    }

    isValid(): boolean {
        return this.hasValidProperties() && this.getFirstDomainFieldError() === undefined;
    }

    isGPAT(): boolean {
        return this.providerName.toLowerCase() === 'general';
    }

    hasValidProperties(): boolean {
        return (
            this.name !== undefined &&
            this.name !== null &&
            this.name.trim().length > 0 &&
            (!this.allowMetadataInputFormatSelection() || Utils.isString(this.selectedMetadataInputFormat)) &&
            (!this.allowDetectionMethodSelection() || Utils.isString(this.selectedDetectionMethod)) &&
            (!this.allowPlateTemplateSelection() || Utils.isString(this.selectedPlateTemplate)) &&
            this.validateTransformScripts() === undefined
        );
    }

    get hasBatchFields(): boolean {
        return this.getDomainByNameSuffix('Batch')?.fields.size > 0;
    }

    getUrl(): AppURL {
        return AppURL.create('assays', this.providerName, this.name);
    }
}
