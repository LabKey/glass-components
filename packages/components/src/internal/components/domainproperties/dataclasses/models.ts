/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Draft, immerable, produce } from 'immer';
import { Map, fromJS, OrderedMap, Set } from 'immutable';

import { DomainDesign, IDomainField, SystemField } from '../models';

import { IParentAlias } from '../../entities/models';

import { DATACLASS_DOMAIN_SYSTEM_FIELDS, SOURCE_DOMAIN_SYSTEM_FIELDS } from './constants';
import {getDuplicateAlias, parentAliasInvalid} from "../utils";

interface DataClassOptionsConfig {
    category: string;
    description: string;
    importAliases?: Map<string, string>;
    name: string;
    nameExpression: string;
    parentAliases?: OrderedMap<string, IParentAlias>;
    rowId: number;
    sampleSet: number;
    systemFields?: SystemField[];
}

export interface DataClassModelConfig extends DataClassOptionsConfig {
    domain: DomainDesign;
    exception: string;
}

export class DataClassModel implements DataClassModelConfig {
    [immerable] = true;

    readonly category: string;
    readonly description: string;
    readonly domain: DomainDesign;
    readonly exception: string;
    readonly name: string;
    readonly nameExpression: string;
    readonly rowId: number;
    readonly sampleSet: number;
    readonly systemFields: SystemField[];
    readonly parentAliases?: OrderedMap<string, IParentAlias>;
    readonly importAliases?: Map<string, string>;
    readonly isBuiltIn?: boolean;

    constructor(values?: Partial<DataClassModelConfig>) {
        Object.assign(this, values);
    }

    static create(raw: any): DataClassModel {
        let domain = DomainDesign.create({});
        if (raw.domainDesign) {
            domain = DomainDesign.create(raw.domainDesign);
        }

        const model = new DataClassModel({ ...(raw.options ? raw.options : raw), domain });

        return produce(model, (draft: Draft<DataClassModel>) => {
            if (raw.options && model.category === null) {
                draft.category = undefined;
            }
            if (raw.options && model.sampleSet === null) {
                draft.sampleSet = undefined;
            }

            const aliases = raw.options?.importAliases || {};
            draft.importAliases = Map<string, string>(fromJS(aliases));

            draft.systemFields =
                model.category === 'sources' ? SOURCE_DOMAIN_SYSTEM_FIELDS : DATACLASS_DOMAIN_SYSTEM_FIELDS;
        });
    }

    get containerPath(): string {
        return this.domain?.container;
    }

    get isNew(): boolean {
        return !this.rowId;
    }

    isValid(defaultNameFieldConfig?: Partial<IDomainField>): boolean {
        return (
            this.hasValidProperties &&
            !this.hasInvalidNameField(defaultNameFieldConfig) &&
            getDuplicateAlias(this.parentAliases, true).size === 0 &&
            !this.domain.hasInvalidFields()
        );
    }

    get hasValidProperties(): boolean {
        const hasInvalidAliases =
            this.parentAliases &&
            this.parentAliases.size > 0 &&
            this.parentAliases.find(parentAliasInvalid) !== undefined;

        return this.name !== undefined && this.name !== null && this.name.trim().length > 0 && !hasInvalidAliases;
    }

    hasInvalidNameField(defaultNameFieldConfig: Partial<IDomainField>): boolean {
        return this.domain && defaultNameFieldConfig ? this.domain.hasInvalidNameField(defaultNameFieldConfig) : false;
    }

    get entityDataMap(): Map<string, any> {
        return fromJS({
            rowId: this.rowId,
            name: this.name,
            description: this.description,
            nameExpression: this.nameExpression,
        });
    }

    get options(): DataClassOptionsConfig {
        return {
            rowId: this.rowId,
            name: this.name,
            description: this.description,
            nameExpression: this.nameExpression,
            category: this.category,
            sampleSet: this.sampleSet,
            systemFields: this.systemFields,
        };
    }

    mutate(props: Partial<DataClassModel>): DataClassModel {
        return produce<DataClassModel>(this, draft => {
            Object.assign(draft, props);
        });
    }
}
