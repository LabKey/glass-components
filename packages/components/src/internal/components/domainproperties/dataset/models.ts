/*
 * Copyright (c) 2020 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Record } from 'immutable';

import { immerable, produce } from 'immer';

import { DomainDesign, DomainField } from '../models';

import { DOMAIN_FIELD_FULLY_LOCKED } from '../constants';

import { allowAsManagedField } from './utils';

export interface DatasetAdvancedSettingsForm {
    cohortId?: number;
    datasetId?: number;
    showByDefault?: boolean;
    tag?: string;
    visitDatePropertyName?: string;
}

export interface IDatasetModel {
    category?: string;
    cohortId?: number;
    dataSharing?: string;
    datasetId?: number;
    definitionIsShared?: boolean;
    demographicData: boolean;
    description?: string;
    domain: DomainDesign;
    domainId: number;
    entityId?: string;
    exception: string;
    keyPropertyManaged: boolean;
    keyPropertyName?: string;
    tag?: string;
    visitDatePropertyName?: string;
    showByDefault: boolean;
    sourceName?: string;
    sourceType?: string;
    name: string;
    sourceUrl?: string;
    label?: string;
    useTimeKeyField?: boolean;
}

export class DatasetModel implements IDatasetModel {
    [immerable] = true;

    readonly domain: DomainDesign;
    readonly domainId: number;
    readonly exception: string;
    readonly datasetId?: number;
    readonly entityId?: string;
    readonly name: string;
    readonly category?: string;
    readonly visitDatePropertyName?: string;
    readonly keyPropertyName?: string;
    readonly keyPropertyManaged: boolean;
    readonly demographicData: boolean;
    readonly label?: string;
    readonly cohortId?: number;
    readonly tag?: string;
    readonly showByDefault: boolean;
    readonly description?: string;
    readonly dataSharing?: string;
    readonly definitionIsShared?: boolean;
    readonly sourceName?: string;
    readonly sourceUrl?: string;
    readonly sourceType?: string;
    readonly useTimeKeyField?: boolean;

    constructor(datasetModel: IDatasetModel) {
        Object.assign(this, datasetModel);
    }

    static create(newDataset = null, raw?: any): DatasetModel {
        if (newDataset) {
            const domain = DomainDesign.create(undefined);
            return new DatasetModel({ ...newDataset, domain });
        } else {
            const domain = DomainDesign.create(raw.domainDesign);
            let model = new DatasetModel({ ...raw.options, domain });

            // if the dataset is from a linked source, disable/lock the fields
            if (model.isFromLinkedSource()) {
                const newDomain = domain.merge({
                    fields: domain.fields
                        .map((field: DomainField) => {
                            return field.set('lockType', DOMAIN_FIELD_FULLY_LOCKED);
                        })
                        .toList(),
                }) as DomainDesign;

                model = produce<DatasetModel>(model, draft => {
                    draft.domain = newDomain;
                });
            }

            return model;
        }
    }

    hasValidProperties(): boolean {
        const isValidName = this.name !== undefined && this.name !== null && this.name.trim().length > 0;
        const isValidLabel =
            this.isNew() || (this.label !== undefined && this.label !== null && this.label.trim().length > 0);

        return isValidName && isValidLabel && this.hasValidAdditionalKey();
    }

    hasValidAdditionalKey(): boolean {
        let isValidKeySetting = true;
        if (this.getDataRowSetting() === 2) {
            isValidKeySetting =
                (this.keyPropertyName !== undefined && this.keyPropertyName !== '') || this.useTimeKeyField;
        }

        return isValidKeySetting;
    }

    isNew(): boolean {
        return !this.entityId;
    }

    getDataRowSetting(): number {
        const noKeyPropName = this.keyPropertyName === undefined || this.keyPropertyName === null;

        // participant id
        if (noKeyPropName && this.demographicData) {
            return 0;
        }
        // participant id and timepoint
        else if (noKeyPropName && !this.useTimeKeyField) {
            return 1;
        }

        // participant id, timepoint and additional key field
        return 2;
    }

    validManagedKeyField(editedName?: string): boolean {
        if (this.keyPropertyName || editedName) {
            const domainFields = this.domain.fields;

            const allowedFieldTypes = domainFields
                .filter(field => allowAsManagedField(field))
                .map(field => {
                    return field.name;
                })
                .toList();

            return allowedFieldTypes.contains(this.keyPropertyName) || allowedFieldTypes.contains(editedName);
        } else {
            return false;
        }
    }

    getDomainKind(timepointType: string): string {
        return timepointType === 'VISIT' ? 'StudyDatasetVisit' : 'StudyDatasetDate';
    }

    getOptions(): Record<string, any> {
        return produce<IDatasetModel>(this, draft => {
            delete draft.exception;
            delete draft.domain;
        });
    }

    isValid(): boolean {
        return this.hasValidProperties() && !this.domain.hasInvalidFields();
    }

    isFromLinkedSource(): boolean {
        return this.sourceName !== undefined && this.sourceName !== null;
    }
}
