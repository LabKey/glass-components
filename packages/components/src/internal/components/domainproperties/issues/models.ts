/*
 * Copyright (c) 2020 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { immerable } from 'immer';

import { DomainDesign } from '../models';

export interface IssuesListDefOptionsConfig {
    entityId?: string;
    issueDefName: string;
    singularItemName: string;
    pluralItemName: string;
    commentSortDirection: string;
    assignedToGroup: number;
    assignedToUser: number;
}

interface IssuesListDefModelConfig extends IssuesListDefOptionsConfig {
    exception: string;
    domain: DomainDesign;
    domainId: number;
    domainKindName: string;
}

export class IssuesListDefModel implements IssuesListDefModelConfig {
    [immerable] = true;

    readonly exception: string;
    readonly domain: DomainDesign;
    readonly entityId?: string;
    readonly domainId: number;
    readonly issueDefName: string;
    readonly singularItemName: string = 'Issue';
    readonly pluralItemName: string = 'Issues';
    readonly commentSortDirection: string = 'ASC';
    readonly assignedToGroup: number;
    readonly assignedToUser: number;
    readonly domainKindName: string;

    constructor(values?: Partial<IssuesListDefModelConfig>) {
        Object.assign(this, values);
    }

    static create(raw: any, defaultSettings = null): IssuesListDefModel {
        if (defaultSettings) {
            const domain = DomainDesign.create(undefined);
            return new IssuesListDefModel({ ...defaultSettings, domain });
        } else {
            const domain = DomainDesign.create(raw.domainDesign);
            return new IssuesListDefModel({ ...raw.options, domain });
        }
    }

    isNew(): boolean {
        return !this.entityId;
    }

    isValid(): boolean {
        return this.hasValidProperties() && !this.domain.hasInvalidFields();
    }

    hasValidProperties(): boolean {
        return this.issueDefName !== undefined && this.issueDefName !== null && this.issueDefName.trim().length > 0;
    }

    getOptions(): IssuesListDefOptionsConfig {
        return {
            entityId: this.entityId,
            issueDefName: this.issueDefName,
            singularItemName: this.singularItemName,
            pluralItemName: this.pluralItemName,
            commentSortDirection: this.commentSortDirection,
            assignedToGroup: this.assignedToGroup,
            assignedToUser: this.assignedToUser,
        };
    }
}
