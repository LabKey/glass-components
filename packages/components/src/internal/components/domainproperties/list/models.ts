/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Record } from 'immutable';

import { DomainDesign, DomainField } from '../models';
import { DOMAIN_FIELD_PRIMARY_KEY_LOCKED } from '../constants';

import { PRIVATE_PICKLIST_CATEGORY, PUBLIC_PICKLIST_CATEGORY } from '../../picklist/constants';

import { INT_LIST, PICKLIST, VAR_LIST } from './constants';

export interface EachItemSettings {
    eachItemBodySetting: number;
    eachItemBodyTemplate: string;
    eachItemTitleTemplate: string;
}

export interface EntireListSettings {
    entireListBodySetting: number;
    entireListBodyTemplate: string;
    entireListIndexSetting: number;
    entireListTitleTemplate: string;
}

export interface AdvancedSettingsForm extends EachItemSettings, EntireListSettings {
    discussionSetting: number;
    eachItemIndex: boolean;
    entireListIndex: boolean;
    fileAttachmentIndex: boolean;
    multiFolder: boolean;
    titleColumn: string;
}

export class ListModel extends Record({
    exception: undefined,
    domain: undefined,
    entityId: undefined,
    createdBy: undefined,
    created: undefined,
    modifiedBy: undefined,
    modified: undefined,
    containerId: undefined,
    name: undefined,
    domainKindName: undefined,
    description: undefined,
    lastIndexed: undefined,
    keyName: undefined,
    titleColumn: undefined,
    domainId: undefined,
    keyType: undefined,
    discussionSetting: undefined,
    allowDelete: undefined,
    allowUpload: undefined,
    allowExport: undefined,
    entireListIndex: undefined,
    entireListIndexSetting: undefined,
    entireListTitleTemplate: undefined,
    entireListBodySetting: undefined,
    entireListBodyTemplate: undefined,
    eachItemIndex: undefined,
    eachItemTitleTemplate: undefined,
    eachItemBodySetting: undefined,
    eachItemBodyTemplate: undefined,
    fileAttachmentIndex: undefined,
    multiFolder: undefined,
    listId: undefined,
    discussionSettingEnum: undefined,
    containerPath: undefined,
    category: undefined,
}) {
    declare exception: string;
    declare domain: DomainDesign;
    declare name: string;
    declare domainKindName: string;
    declare description: string;
    declare lastIndexed: any;
    declare keyName: string;
    declare titleColumn: string;
    declare domainId: number;
    declare keyType: string;
    declare discussionSetting: number;
    declare allowDelete: true;
    declare allowUpload: true;
    declare allowExport: true;
    declare entireListIndex: true;
    declare entireListIndexSetting: number;
    declare entireListTitleTemplate: string;
    declare entireListBodySetting: number;
    declare entireListBodyTemplate: string;
    declare eachItemIndex: false;
    declare eachItemTitleTemplate: string;
    declare eachItemBodySetting: number;
    declare eachItemBodyTemplate: string;
    declare fileAttachmentIndex: false;
    declare multiFolder: false;
    declare listId: number;
    declare discussionSettingEnum: string;
    declare containerPath: string;
    declare category: string;

    static create(raw: any, defaultSettings = null): ListModel {
        if (defaultSettings) {
            const domain = DomainDesign.create(undefined);
            return new ListModel({ ...defaultSettings, domain });
        } else {
            let domain = DomainDesign.create(raw.domainDesign);

            // Issue39818: Set the key field of an existing list to be PKLocked.
            const fields = domain.fields;
            const pkField = fields.findIndex(i => i.isPrimaryKey);
            if (pkField > -1) {
                const pkFieldLocked = fields
                    .get(pkField)
                    .merge({ lockType: DOMAIN_FIELD_PRIMARY_KEY_LOCKED }) as DomainField;
                const updatedFields = fields.set(pkField, pkFieldLocked);
                domain = domain.set('fields', updatedFields) as DomainDesign;
            }

            return new ListModel({ ...raw.options, domain });
        }
    }

    getDomainKind(): string {
        if (this.category === PUBLIC_PICKLIST_CATEGORY || this.category === PRIVATE_PICKLIST_CATEGORY) {
            return PICKLIST;
        } else if (this.keyType === 'Varchar') {
            return VAR_LIST;
        } else if (this.keyType === 'Integer' || this.keyType === 'AutoIncrementInteger') {
            return INT_LIST;
        }

        return undefined;
    }

    hasValidKeyType(): boolean {
        return this.getDomainKind() !== undefined;
    }

    isNew(): boolean {
        return !this.listId;
    }

    isValid(): boolean {
        return this.hasValidProperties() && this.hasValidKeyType() && !this.domain.hasInvalidFields();
    }

    hasValidProperties(): boolean {
        return this.name !== undefined && this.name !== null && this.name.trim().length > 0;
    }

    getOptions(): Record<string, any> {
        const options = this.toJS();

        // Note: keyName is primarily set using <SetKeyFieldNamePanel/>'s onSelectionChange()
        // Setting keyName here covers the use-case where a user sets a Key Field, and then changes its name
        const keyField = this.domain.fields.find(field => field.isPrimaryKey);
        if (keyField && this.keyName !== keyField.name) {
            options.keyName = keyField.name;
        }

        delete options.exception;
        delete options.domain;
        return options;
    }
}
