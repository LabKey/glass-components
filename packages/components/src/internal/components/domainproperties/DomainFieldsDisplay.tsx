/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React, { FC } from 'react';

import { GridColumn } from '../base/models/GridColumn';
import { Grid } from '../base/Grid';

import { DomainDesign } from './models';

export const DOMAIN_FIELD_COLS = [
    new GridColumn({ index: 'name', title: 'Name' }),
    new GridColumn({ index: 'label', title: 'Label' }),
    new GridColumn({ index: 'rangeURI', title: 'Range URI' }),
    new GridColumn({ index: 'conceptURI', title: 'Concept URI' }),
    new GridColumn({ index: 'required', title: 'Required' }),
    new GridColumn({ index: 'scale', title: 'Scale' }),
];

type Props = {
    domain: DomainDesign;
    title?: string;
};

export const DomainFieldsDisplay: FC<Props> = ({ domain, title }) => (
    <div className="panel panel-default">
        <div className="panel-heading">
            <div className="panel-title">{title || domain.name}</div>
        </div>
        <div className="panel-body">
            <p>{domain.description}</p>
            <Grid columns={DOMAIN_FIELD_COLS} data={domain.fields} />
        </div>
    </div>
);
DomainFieldsDisplay.displayName = 'DomainFieldsDisplay';
