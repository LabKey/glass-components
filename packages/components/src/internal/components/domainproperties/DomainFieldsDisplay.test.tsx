/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { render } from '@testing-library/react';

import { DomainFieldsDisplay } from './DomainFieldsDisplay';
import { DomainDesign } from './models';

const testDomain = new DomainDesign({ name: 'test domain name' });

describe('DomainFieldsDisplay', () => {
    test('with empty domain design', () => {
        const domain = new DomainDesign();
        const { container} = render(<DomainFieldsDisplay domain={domain} />);

        expect(container.firstChild).toMatchSnapshot();
    });

    test('without title', () => {
        const { container} = render(<DomainFieldsDisplay domain={testDomain} />);

        expect(container.firstChild).toMatchSnapshot();
    });

    test('with title', () => {
        const { container} = render(<DomainFieldsDisplay domain={testDomain} title="test domain title" />);

        expect(container.firstChild).toMatchSnapshot();
    });
});
