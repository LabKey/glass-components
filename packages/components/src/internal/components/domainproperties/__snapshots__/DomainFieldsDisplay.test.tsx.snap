// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DomainFieldsDisplay with empty domain design 1`] = `
<div>
  <div
    class="panel panel-default"
  >
    <div
      class="panel-heading"
    >
      <div
        class="panel-title"
      />
    </div>
    <div
      class="panel-body"
    >
      <p />
      <div
        class="table-responsive"
      >
        <table
          class="table table-striped table-bordered"
        >
          <thead>
            <tr>
              <th
                class="grid-header-cell"
                draggable="false"
                id="name"
              >
                Name
              </th>
              <th
                class="grid-header-cell"
                draggable="false"
                id="label"
              >
                Label
              </th>
              <th
                class="grid-header-cell"
                draggable="false"
                id="rangeURI"
              >
                Range URI
              </th>
              <th
                class="grid-header-cell"
                draggable="false"
                id="conceptURI"
              >
                Concept URI
              </th>
              <th
                class="grid-header-cell"
                draggable="false"
                id="required"
              >
                Required
              </th>
              <th
                class="grid-header-cell"
                draggable="false"
                id="scale"
              >
                Scale
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              class="grid-empty"
            >
              <td
                colspan="6"
              >
                No data available.
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`DomainFieldsDisplay with title 1`] = `
<div>
  <div
    class="panel panel-default"
  >
    <div
      class="panel-heading"
    >
      <div
        class="panel-title"
      >
        test domain title
      </div>
    </div>
    <div
      class="panel-body"
    >
      <p />
      <div
        class="table-responsive"
      >
        <table
          class="table table-striped table-bordered"
        >
          <thead>
            <tr>
              <th
                class="grid-header-cell"
                draggable="false"
                id="name"
              >
                Name
              </th>
              <th
                class="grid-header-cell"
                draggable="false"
                id="label"
              >
                Label
              </th>
              <th
                class="grid-header-cell"
                draggable="false"
                id="rangeURI"
              >
                Range URI
              </th>
              <th
                class="grid-header-cell"
                draggable="false"
                id="conceptURI"
              >
                Concept URI
              </th>
              <th
                class="grid-header-cell"
                draggable="false"
                id="required"
              >
                Required
              </th>
              <th
                class="grid-header-cell"
                draggable="false"
                id="scale"
              >
                Scale
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              class="grid-empty"
            >
              <td
                colspan="6"
              >
                No data available.
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`DomainFieldsDisplay without title 1`] = `
<div>
  <div
    class="panel panel-default"
  >
    <div
      class="panel-heading"
    >
      <div
        class="panel-title"
      >
        test domain name
      </div>
    </div>
    <div
      class="panel-body"
    >
      <p />
      <div
        class="table-responsive"
      >
        <table
          class="table table-striped table-bordered"
        >
          <thead>
            <tr>
              <th
                class="grid-header-cell"
                draggable="false"
                id="name"
              >
                Name
              </th>
              <th
                class="grid-header-cell"
                draggable="false"
                id="label"
              >
                Label
              </th>
              <th
                class="grid-header-cell"
                draggable="false"
                id="rangeURI"
              >
                Range URI
              </th>
              <th
                class="grid-header-cell"
                draggable="false"
                id="conceptURI"
              >
                Concept URI
              </th>
              <th
                class="grid-header-cell"
                draggable="false"
                id="required"
              >
                Required
              </th>
              <th
                class="grid-header-cell"
                draggable="false"
                id="scale"
              >
                Scale
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              class="grid-empty"
            >
              <td
                colspan="6"
              >
                No data available.
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
